mdsuite/calculators/structure_factor.py

```diff
@@ -99,246 +110,264 @@ class StructureFactor(Calculator):
+            partial_structure_factors[pair] = integral
+
+        return partial_structure_factors
+
+    def _compute_weight_factors(self) -> dict:
         """
-        molar_fractions = {}
+        Compute the weight factors for the SF computation.
 
-        for el in self.experiment.species:
-            molar_fractions[el] = (
-                self.experiment.species[el].n_particles / self.experiment.number_of_atoms
+        Compute the weight factors from the atomic form factors (aff) for each species.
+
+        Returns
+        -------
+        weight_factors : dict
+                A dict of weight factors to be used in the SF computation. There is
+                one weight factor for each pair.
+        """
+        weight_factors = {}
+        for pair, pair_data in self.rdf_data.data_dict.items():
+            species_names = pair.split("_")
+            form_factors = [self.species_dict[item].form_factor for item in species_names]
+            mean_square_form_factor = np.mean(form_factors) ** 2
+            molar_fraction = np.prod(
+                [self.species_dict[item].molar_fraction for item in species_names]
+            )
+            weight_factors[pair] = (
+                molar_fraction * np.prod(form_factors) / mean_square_form_factor
             )
 
-        return molar_fractions
+        return weight_factors
 
-    @property
-    def species_densities(self):
-        """Calculates the particle densities
+    def _compute_total_structure_factor(
+        self, partial_sf: dict, weight_factors: dict
+    ) -> np.ndarray:
+        """
+        Compute the total structure factor.
 
-        Calculates the particle densities for all the species in the species
-        dictionary and add it to the species dictionary
+        Parameters
+        ----------
+        partial_sf : dict
+                A dict of partial structure factors for each pair.
+        weight_factors : dict
+                A dict of weight factors computed using the relevant formalism for
+                each pair.
 
-        # TODO this is uncached!!
-        # TODO was species[x]["species_densities"] ever used somewhere?
-        #   this needs tests!
+        Returns
+        -------
+        total_structure_factor : np.ndarray
+                Total structure factor of the system.
         """
-        species_densities = {}
-        for el in self.experiment.species:
-            species_densities[el] = self.experiment.species[el].n_particles / (
-                self.experiment.box_array[0]
-                * self.experiment.box_array[1]
-                * self.experiment.box_array[2]
-            )
-        return species_densities
+        structure_factor = np.zeros(self.args.resolution)
+        for pair, pair_data in partial_sf.items():
+            species = pair.split("_")
+            if species[0] == species[1]:
+                factor = 2
+            else:
+                factor = 1
```

Comment:
Confuse people with:
`factor = int(species[0] == species[1]) + 1`

Reply:
I love it

Reply:
@christophlohrmann suggested `species = 2 if species[0] == species[1] else 1`. I don't understand why anyone could dislike casting boolean to integer.

Reply:
In general I think keeping one-liners that aren't extremely transparent out of code. I dont always think the shortes solution is the best when it comes to code clarity, particularly is performance is not impacted.